rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,518 @@
+            "/ids/{role}/current": {
+                "kind": "identity",
+                "version": "v2",
+                "metadata": {
+                    "name": "current"
+                },
+                "spec": {
+                    "key": "{key_content}",
+                    "ssh_cert": "{ssh_cert_content}",
+                    "tls_cert": "{tls_cert_content}",
+                    "tls_ca_certs": ["{tls_ca_certs}"],
+                    "ssh_ca_certs": ["{ssh_ca_certs}"]
+                }
+            }
+
+             # State is important if restart/rollback happens during the CA rotation phase.
+             # it holds the current status of the rotation
+            "/states/{role}/state": {
+                "kind": "state",
+                "version": "v2",
+                ...
+            }
+            
+            # during CA rotation phase, the new keys are stored if agent is restarted or rotation has to rollback
+            "/ids/{role}/replacement": {
+                "kind": "identity",
+                "version": "v2",
+                "metadata": {
+                    "name": "replacement"
+                },
+                "spec": {
+                    "key": "{key_content}",
+                    "ssh_cert": "{ssh_cert_content}",
+                    "tls_cert": "{tls_cert_content}",
+                    "tls_ca_certs": ["{tls_ca_certs}"],
+                    "ssh_ca_certs": ["{ssh_ca_certs}"]
+                }
+            }
+        }
+    
+kind: Secret
+metadata:
+  name: {.Release.Name}-identity-{{$TELEPORT_REPLICA_NAME}}
+  namespace: {.Release.Namespace}
+```
+
+Where:
+
+- `ssh_cert` is a PEM encoded SSH host cert.
+- `key` is a PEM encoded private key.
+- `tls_cert` is a PEM encoded x509 client certificate.
+- `tls_ca_certs` is a list of PEM encoded x509 certificate of the certificate authority of the cluster.
+- `ssh_ca_certs` is a list of SSH certificate authorities encoded in the authorized_keys format.
+- `role` is the role the agent is operating. If agent is running with multiple roles, i.e. app, kube..., multiple entries will be added, one for each role.
+- `TELEPORT_REPLICA_NAME` is the teleport agent replica name. Constant when using Deployments, `TELEPORT_REPLICA_NAME={{ .Release.Name}}` or dynamic when using Statefulsets, `TELEPORT_REPLICA_NAME=metadata.name`.
+
+#### RBAC Changes
+
+The Teleport agent service account must be able to create, read and update secrets within the namespace, therefore, one must create a new namespace role and attach it to the agent service account with the following content:
+
+```yaml
+- apiGroups: [""]
+  # objects is "secrets"
+  resources: ["secrets"]
+  verbs: ["create"] # create must have a special case in RBAC since we cannot block creation of the secret based on resource name.
+- apiGroups: [""]
+  # objects is "secrets"
+  resources: ["secrets"]
+  resourcesNames: 
+  - "{.Release.Name}-identity-{{$TELEPORT_REPLICA_NAME}}" # initial secrets
```

Comment:
We know the `TELEPORT_REPLICA_NAME` beforehand.

- Deployment: {{.Release.Name}} // fixed 
- Statefulset: {{.Release.Name}}-0, {{.Release.Name}}-1, {{.Release.Name}}-2...

Although allowing it to operate on secrets within the namespace would allow the users to basically scale it whenever they want to without requiring the usage of Helm. This was my initial thought but if it's not installed in a separate namespace might be a bit more problematic.